pkg/gadgets/trace/tcpdrop/tracer/bpf/tcpdrop.bpf.c

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: GPL-2.0
+//
+// Based on tcpdrop(8) from BCC
+//
+// Copyright 2018 Netflix, Inc.
+// 30-May-2018    Brendan Gregg   Created this.
+// 15-Jun-2022    Rong Tao        Add tracepoint:skb:kfree_skb
+// Copyright 2023 Microsoft Corporation
+
+#include <vmlinux/vmlinux.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_endian.h>
+
+#include "tcpdrop.h"
+
+/* Define here, because there are conflicts with include files */
+#define AF_INET		2
+#define AF_INET6	10
+
+// we need this to make sure the compiler doesn't remove our struct
+const struct event *unusedevent __attribute__((unused));
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+// This struct is defined according to the following format file:
+// /sys/kernel/tracing/events/skb/kfree_skb/format
+struct kfree_skb_ctx {
+	/* The first 8 bytes is not allowed to read */
+	unsigned long pad;
+
+	void *skbaddr;
+	void *location;
+	unsigned short protocol;
+	int reason;
+};
+
+static __always_inline int __trace_tcp_drop(void *ctx, struct sock *sk, struct sk_buff *skb, int reason)
+{
+	if (sk == NULL)
+		return 0;
+
+	u_int8_t *tcphdr = (u_int8_t *)(BPF_CORE_READ(skb, head) + BPF_CORE_READ(skb, transport_header));
+	struct inet_sock *sockp = (struct inet_sock *)sk;
+	struct task_struct *task = (struct task_struct*) bpf_get_current_task();
+
+	struct event event = {};
+	event.timestamp = bpf_ktime_get_boot_ns();
+	event.af = BPF_CORE_READ(sk, __sk_common.skc_family);
+	event.state = BPF_CORE_READ(sk, __sk_common.skc_state);
+	event.reason = reason;
+	bpf_get_current_comm(&event.comm, sizeof(event.comm));
+	event.pid = bpf_get_current_pid_tgid() >> 32;
+	event.mntns_id = (u64) BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
```

Comment:
Missing logic to filter by mntns id?

Reply:
I removed that logic because the tcpdrop events might be triggered in the kernel from interruption context, or from processes unrelated to the process sending/receiving the packets...

I was wondering about hiding the container/process columns by default to avoid confusing the user.

Example of output where the container/process is not meaningful:
```
CONTAINER                                                    PID        COMM             IP SADDR                            DADDR                            SPORT DPORT STATE        TCPFLAGS                        REASON                 
                                                             1007       irq/130-iwlwifi  4  172.17.0.2                       1.1.1.1                          53612 443   ESTABLISHED  ACK                             QDISC_DROP             
```

Reply:
I think we should at least hide the container and k8s related ones. I'm wondering if we can use the socket enricher in this gadget too?

Reply:
> the tcpdrop events might be triggered in the kernel from interruption context, or from processes unrelated to the process sending/receiving the packets...
>

Mmmm are all these kinds of events triggered twice like the following?

```bash
CONTAINER  PID        COMM       IP SADDR       DADDR    SPORT DPORT STATE        TCPFLAGS  REASON
netem      3807012    wget       4  172.17.0.2  1.1.1.1  54070 80    ESTABLISHED  PSH|ACK   QDISC_DROP
           0          swapper/2  4  172.17.0.2  1.1.1.1  54070 80    ESTABLISHED  PSH|ACK   QDISC_DROP
```

If so, is it possible to drop them and just keep the ones happening in the container?

Reply:
> I think we should at least hide the container and k8s related ones.

Ok, I hide them. Users can add them back with `-o columns=container,comm,saddr,daddr,state,tcpflags,reason`

> I'm wondering if we can use the socket enricher in this gadget too?

It might be possible, but not trivial:
- We cannot use directly `gadget_socket_lookup()` because we don't have the `struct __sk_buff` (from networking ebpf programs) but only the internal `struct sk_buff` (from tracing ebpf programs). So we need another implementation.
- I don't know if we have the netns from the `struct sk_buff` because the skb is traveling in different netns and the current netns is not relevant here. We might need to lookup in all network namespaces.

I added this as TODO in the description for a future PR.

> Mmmm are all these kinds of events triggered twice like the following?

Not necessarily. In this scenario, we drop randomly 20% of the packets. Sometimes the random generator selects only packets that happens to be in interruption context (`irq/130-iwlwifi`) so we don't have any container information.